#!/usr/bin/env Rscript

# @file step2_1_calculate_PEER_residuals
# @title Calculate PEER Residuals
# @description This script calculates PEER residuals for gene expression data by performing linear regression, adjusting for covariates and PEER factors, and then outputs the residuals.
# @param expr_file Path to the log2 transformed z-normalized expression matrix file.
# @param covs_file Path to the covariates file.
# @param peer_file Path to the PEER factors file generated by calculate_PEER_factors.R.
# @param out_file Output file path where the residuals will be saved.
# @param num_pcs Number of genotype PCs to retain; defaults to 5.
# @usage Rscript step2_1_calculate_PEER_residuals.R --expr=LOGZ_EXPR_FILE --cov=COV_FILE --peer=PEER_FILE --out=OUT_FILE --pcs=5
# @return Outputs a file containing the residuals.
# @examples 
# Rscript step2_1_calculate_PEER_residuals.R --expr="Kidney_Cortex.log2.ztrans.txt" --cov="combined_covariates" --peer="factors.tsv" --out="Kidney_Cortex_residuals.txt" --pcs=5

library(optparse)
library(data.table)
library(plyr)
library(dplyr)

# Option handling
option_list = list(
  make_option(c("--expr"), type = "character", help = "Path to the expression file"),
  make_option(c("--cov"), type = "character", help = "Path to the covariate file"),
  make_option(c("--peer"), type = "character", help = "Path to the PEER factors file"),
  make_option(c("--out"), type = "character", help = "Output file name"),
  make_option(c("--pcs"), type = "integer", help = "Number of genotype PCs to retain", default = 5)
)

opt_parser = OptionParser(option_list=option_list)
options = parse_args(opt_parser)

process_data <- function(expr_file, covs_file, peer_file, out_file, num_pcs) {
  ## Read in expression and covariate matrices
  expr = read.table(expr_file, header = T, sep = '\t', row.names = 1)
  covs = read.table(covs_file, header = T, sep = '\t', row.names = 1)
  peer_factors = t(read.table(peer_file, header = T, sep = '\t', row.names = 1))

  # Read data
  #expr = fread(expr_file, data.table = FALSE)
  #covs = fread(covs_file, data.table = FALSE)
  #peer_factors = t(fread(peer_file, data.table = FALSE))
  
  # Adjust data
  rownames(peer_factors) = gsub("\\.", "-", rownames(peer_factors))
  covs = covs[rownames(expr), ]
  covs = covs[, !(colnames(covs) %in% paste0('PC', (num_pcs + 1):20))]
  peer_factors = peer_factors[rownames(expr), ]
  covs = cbind(covs, peer_factors)
  
  # Remove NA
  inds_to_keep = rowSums(is.na(covs)) == 0
  covs = covs[inds_to_keep, ]
  expr = expr[inds_to_keep, ]
  
# Linear regression per gene
  resids = matrix(, ncol = ncol(expr), nrow = nrow(expr))
  rownames(resids) = rownames(expr)
  colnames(resids) = colnames(expr)
  
  for(i in 1:ncol(expr)){
	#print(i)
	gene = names(expr)[i]
	data = as.data.frame(cbind(expr[, i], covs))
	colnames(data) = c('RPKM', colnames(covs))
	model = lm(RPKM ~ ., data = data)
	resids[, i] = model$residuals
  }
  
  # Write out residual results
  resids = t(scale(resids))
  write.table(matrix(c('Id', colnames(resids)), nrow = 1), out_file, quote = F, row.names = F, col.names = F, sep = '\t')
  write.table(resids, out_file, row.names = T, col.names = F, quote = F, sep = '\t', append = T)
  #write.table(cbind('Id', colnames(expr)), out_file, quote = FALSE, row.names = FALSE, col.names = FALSE, sep = '\t')
  #write.table(resids, out_file, row.names = TRUE, col.names = FALSE, quote = FALSE, sep = '\t', append = TRUE)
}

# Run the processing function
process_data(options$expr, options$cov, options$peer, options$out, options$pcs)

